<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>tess3Main. tess3r</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>

<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container">
      <header>
      
<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">tess3r</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/main-vignette.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/BioShock38/TESS3_encho_sen">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
      
      </header>

      <div class="page-header">
  <h1>Estimates spatial population structure</h1>
</div>

<div class="row">
  <div class="col-md-9">
    
    <p><code>tess3Main</code> estimates spatial population structure using a graph based non
negative matrix factorization. After estimating the population structure is used to
compute a Fst statistic for each locus. See references for more details.</p>
    

    <pre><span class='fu'>tess3Main</span>(<span class='no'>X</span>, <span class='kw'>XProba</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>coord</span>, <span class='no'>K</span>, <span class='no'>ploidy</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>, <span class='kw'>max.iteration</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>tolerance</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>openMP.core.num</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>Q.init</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mask</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>copy</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>algo.copy</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2>Arguments</h2>
    <dl class="dl-horizontal">
      <dt>X</dt>
      <dd>A numeric matrix which corresponds to the genotype matrix. This matrix
must be of size \(n \times L\) where \(n\) is the number of individual and
\(L\) is the number of loci. Values of this matrix are integer corresponding
to the number of variant alleles observed at a locus. If <code>NULL</code>, <code>XProba</code>
is used.</dd>
      <dt>XProba</dt>
      <dd>A numeric matrix which correspond to the probability for the genotype.
This matrix must be of size \(n \times (ploidy + 1)L\) where
\(n\) is the number of individual, \(L\) is the number of loci. Values of
this matrix are numeric between 0 and 1 corresponding
to the genome probability. It is the matrix used in graph based non negative
factorization matrix. If <code>NULL</code>, it is computed from the genotype matrix <code>X</code>.
See reference for more details.</dd>
      <dt>coord</dt>
      <dd>The numeric matrix of size \(n \times 2\) where \(n\) is the
number of individuals.</dd>
      <dt>K</dt>
      <dd>An integer which corresponds to
the number of ancestral populations.</dd>
      <dt>ploidy</dt>
      <dd>An integer which corresponds to
the number of set of chromosomes.</dd>
      <dt>lambda</dt>
      <dd>A numeric which corresponds to the
spatial regularization parameter.</dd>
      <dt>W</dt>
      <dd>A numeric matrix which corresponds to the graph weiht matrix.
If NULL, W it is computed as
<code>W[i,j] = exp( - (coord[i] - coord[j])^2 / sigma^2)</code>.
Where <code>coord[i]</code> is
the geographic coordinate for the individual i and
<code>sigma</code> equals 5 percent of the average geographic distance between individuals.</dd>
      <dt>method</dt>
      <dd><code>&quot;projected.ls&quot;</code> or <code>&quot;qp&quot;</code>. If <code>&quot;projected.ls&quot;</code>,
an aleternated projected least squares algorithm is used. If <code>&quot;qp&quot;</code>,
an alternated quadratic programing algorithm is used. See references for more
details</dd>
      <dt>max.iteration</dt>
      <dd>The max number of
iteration of the optimization algorithm.</dd>
      <dt>tolerance</dt>
      <dd>A numeric which corresponds to the tolerance of the
stopping criteria of the optimization algorithm.</dd>
      <dt>openMP.core.num</dt>
      <dd>If openMP is available on your computer, it is the
number of core used by the algorithm.</dd>
      <dt>Q.init</dt>
      <dd>A numeric matrix which corresponds to the initial value of
<code>Q</code> for the algorithm.</dd>
      <dt>mask</dt>
      <dd>If not <code>NULL</code> this the proportion of the data matrix which
is masked to compute the cross validation criteria.</dd>
      <dt>copy</dt>
      <dd>if TRUE data will be copy 1 time.</dd>
      <dt>algo.copy</dt>
      <dd>if TRUE data will be copy 1 time to speed the algorithm.</dd>
      <dt>verbose</dt>
      <dd>If <code>TRUE</code> more information are printed.</dd>
    </dl>
    
    <div class="Value">
      <h2>Value</h2>

      <p>An object of class tess3Main which is a list with components:
<dl class='dl-horizontal'>
   <dt>L</dt><dd>The number of loci.</dd>
   <dt>n</dt><dd>The number of individuals.</dd>
   <dt>ploidy</dt><dd>The number of set of chromosomes.</dd>
   <dt>K</dt><dd>The number of ancestral population.</dd>
   <dt>G</dt><dd>The ancestral genotype frequency matrix.</dd>
   <dt>Q</dt><dd>The ancestry coefficient matrix.</dd>
   <dt>Fst</dt><dd>The Fst statistic computed for each locus.</dd>
   <dt>Fscore</dt><dd>The Fscore computed from Fst.</dd>
   <dt>pvalue</dt><dd>The pvalues computed from Fscore.</dd>
   <dt>log.pvalue</dt><dd>The \(log(pvalue)\).</dd>
   <dt>rmse</dt><dd>The root square mean error between <code>XProba</code> and
   <code>tcrossprod(Q, G)</code>.</dd>
   <dt>crossentropy</dt><dd>The cross entropy error between <code>XProba</code> and
   <code>tcrossprod(Q, G)</code>.</dd>
  <dt>crossvalid.rmse</dt><dd>If masked not <code>NULL</code>.
  The root square mean error between <code>XProba[masked]</code> and
  <code>tcrossprod(Q, G)[masked]</code>.</dd>
   <dt>crossvalid.crossentropy</dt><dd>If masked not <code>NULL</code>.
   The cross entropy error between <code>XProba[masked]</code> and
   <code>tcrossprod(Q, G)[masked]</code>.</dd>
</dl></p>
    </div>

    <div class="References">
      <h2>References</h2>

      <p><a href = 'https://hal.archives-ouvertes.fr/hal-01222555/'>https://hal.archives-ouvertes.fr/hal-01222555/</a> <a href = 'http://biorxiv.org/content/early/2016/10/12/080291'>http://biorxiv.org/content/early/2016/10/12/080291</a></p>
    </div>

    <div class="See also">
      <h2>See also</h2>

      <p><code><a href='tess3.html'>tess3</a></code></p>
    </div>
    
    <h2 id="examples">Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>tess3r</span>)

<span class='co'># Arabidopsis thaliana data set</span>
<span class='fu'>data</span>(<span class='no'>data.at</span>)
<span class='no'>genotype</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>X</span>
<span class='no'>coordinates</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>coord</span>

<span class='co'># Run of tess3 main algorithm</span>
<span class='no'>tess3.obj</span> <span class='kw'>&lt;-</span> <span class='fu'>tess3Main</span>(<span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>genotype</span>,
                      <span class='kw'>coord</span> <span class='kw'>=</span> <span class='no'>coordinates</span>,
                      <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>,
                      <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>,
                      <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>1</span>)</div><div class='output co'>#&gt; Computing spectral decomposition of graph laplacian matrix: done
#&gt; Main loop with 1 threads: 
#&gt; : done
#&gt; </div><div class='input'>
<span class='co'># Run of tess3 main algorithm with cross validation errors computation.</span>
<span class='no'>tess3.obj</span> <span class='kw'>&lt;-</span> <span class='fu'>tess3Main</span>(<span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>genotype</span>,
                      <span class='kw'>coord</span> <span class='kw'>=</span> <span class='no'>coordinates</span>,
                      <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>,
                      <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>,
                      <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                      <span class='kw'>mask</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)</div><div class='output'><strong class='text-info'>Mask 0.05% of genotypes for cross validation</strong></div><div class='output'><strong class='text-info'>Missing value detected in genotype</strong></div><div class='output co'>#&gt; Computing spectral decomposition of graph laplacian matrix: done
#&gt; Main loop with 1 threads: 
#&gt; : done
#&gt; </div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3">
  </div>
</div>

      <footer>
      <p>Built by <a href="http://hadley.github.io/pkgdown/">pkgdown</a>. Styled with <a href="http://getbootstrap.com">Bootstrap 3</a>.</p>
      </footer>
   </div>

  </body>
</html>
