<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Estimate ancestry coefficients and run genome scans for selection — tess3 • tess3r</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">tess3r</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/main-vignette.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bcm-uga/TESS3_encho_sen">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate ancestry coefficients and run genome scans for selection</h1>
    </div>

    
    <p><code>tess3</code> is the main function of the <code>tess3r</code> package. It runs
a graph-based nonnegative matrix factorization algorithm that includes
geographic data in the estimation of spatial population structure.
The function requires individual genotypes, geographic coordinates, and
it can be run for multiple values of the number of ancestral populations.
In addition, the function uses the estimates of ancestry coefficients to
compute an Fst statistic for each locus, and to return test significance
values for a null hypothesis of selective neutrality. See the references
for more details.</p>
    

    <pre class="usage"><span class='fu'>tess3</span>(<span class='no'>X</span>, <span class='kw'>XProba</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>coord</span>, <span class='no'>K</span>, <span class='no'>ploidy</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>rep</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>, <span class='kw'>max.iteration</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>tolerance</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>openMP.core.num</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>Q.init</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mask</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>algo.copy</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>keep</span> <span class='kw'>=</span> <span class='st'>"best"</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>X</th>
      <td><p>a matrix of individual genotypes. This matrix must
have \(n\) rows and  \(L\) columns where \(n\) is the number of individuals and
\(L\) is the number of loci. The entries of this matrix are integers between 0 and
ploidy, that correspond to the number of derived/reference alleles observed at each locus.
If <code>NULL</code>, the matrix of genotype likelihood, <code>XProba</code>, is used.</p></td>
    </tr>
    <tr>
      <th>XProba</th>
      <td><p>A matrix which contains individual genotype likelihoods (probabilities) for each
locus. This matrix must contain \(n\) rows and \((ploidy + 1)L\) columns where
\(n\) is the number of individuals, and \(L\) is the number of loci. The entries of
this matrix are numeric values ranging between 0 and 1, and corresponding to genotype
probabilities for each locus. If <code>NULL</code>, deterministic values are computed from the
genotypic matrix <code>X</code>. See the references for more details.</p></td>
    </tr>
    <tr>
      <th>coord</th>
      <td><p>a matrix of size \(n \times 2\) where \(n\) is the
number of individuals encoding the longitude and latitude of each individual (numeric system).</p></td>
    </tr>
    <tr>
      <th>K</th>
      <td><p>an integer or a vector of integers corresponding to
the number of ancestral populations.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>an integer corresponding to ploidy of the studied
organism. Haploids have ploidy = 1, diploids have ploidy = 2, etc.</p></td>
    </tr>
    <tr>
      <th>lambda</th>
      <td><p>a numeric value for the spatial regularization parameter.
The default value lambda = 1 attributes equal weights to the loss function
and to the penalty function.</p></td>
    </tr>
    <tr>
      <th>rep</th>
      <td><p>integer. The number of time the algorithm will be repeated for each value of
<code>K</code>.</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>a matrix which corresponds to the graph weightings.
If NULL, W is computed as
<code>W[i,j] = exp(-(coord[i] - coord[j])^2 / sigma^2)</code>,
where <code>coord[i]</code> represents
the geographic coordinates for individual i, and where
<code>sigma</code> is equal to 5 percent of the average geographic distance between individuals.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>a character string <code>"projected.ls"</code> or <code>"qp"</code>. If <code>"projected.ls"</code>,
an alternating projected least squares algorithm is used. If <code>"qp"</code>,
an alternating quadratic programing algorithm is used. See references for more
details</p></td>
    </tr>
    <tr>
      <th>max.iteration</th>
      <td><p>the maximum number of iterations of the optimization algorithm.</p></td>
    </tr>
    <tr>
      <th>tolerance</th>
      <td><p>a numeric value corresponding to the
stopping criteria of the optimization algorithm.</p></td>
    </tr>
    <tr>
      <th>openMP.core.num</th>
      <td><p>integer representing the number of cores used by the optimization
algorithm. It requires that the openMP library is installed in your OS (default for macOS is no).</p></td>
    </tr>
    <tr>
      <th>Q.init</th>
      <td><p>a matrix for initial values of ancestry coefficients for the algorithm. The
default value is a random matrix.</p></td>
    </tr>
    <tr>
      <th>mask</th>
      <td><p>If not <code>NULL</code>, this numeric value is the proportion of masked data
when computing the cross-validation criterion.</p></td>
    </tr>
    <tr>
      <th>algo.copy</th>
      <td><p>boolean. If TRUE data is copied to speed up the algorithm.</p></td>
    </tr>
    <tr>
      <th>keep</th>
      <td><p>If <code>"best"</code>, only the result with the lowest <code>rmse</code> score will be kept
for each value of <code>K</code>. If <code>"all"</code>, all results will be kept
and returned for each value of <code>K</code>. The second option uses more space in memory.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>If <code>TRUE</code> more information is printed.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class tess3 which corresponds to a list of length <code>length(K)</code>.
Each element of this list has the following attributes</p><dl class='dl-horizontal'>
   <dt>K</dt><dd><p>the number of ancestral populations</p></dd>
   <dt>tess3.run</dt><dd><p>if <code>keep = "best"</code>, the <code><a href='tess3Main.html'>tess3Main</a></code> result
   with the lowest value of the <code>rmse</code> (loss) function. If <code>keep = "all"</code>,
   a list of <code><a href='tess3Main.html'>tess3Main</a></code> results for each repetition</p></dd>
   <dt>rmse</dt><dd><p>root mean squared error between the genotypic matrix <code>XProba</code> and the
   fitted matrix for each program repetition</p></dd>
   <dt>crossentropy</dt><dd><p>cross-entropy between the genotypic matrix <code>XProba</code> and the
   fitted matrix for each program repetition</p></dd>
   <dt>crossvalid.rmse</dt><dd><p>root square mean error between the masked values of genotypic matrix
    <code>XProba[masked]</code> and their fitted values for each repetition. If mask is FALSE, then <code>NULL</code>.</p></dd>
   <dt>crossvalid.crossentropy</dt><dd><p>cross-entropy between the masked values of genotypic matrix
    <code>XProba[masked]</code> and their fitted values for each repetition. If mask is FALSE, then <code>NULL</code>.</p></dd>
</dl><p>Methods available for this class:</p><ul>
<li><p><code><a href='plot.tess3.html'>plot.tess3</a></code></p></li>
<li><p><code><a href='summary.tess3.html'>summary.tess3</a></code></p></li>
<li><p><code><a href='is.tess3.html'>is.tess3</a></code></p></li>
<li><p><code><a href='Gettess3res.html'>Gettess3res</a></code></p></li>
<li><p><code><a href='qmatrix.html'>qmatrix</a></code></p></li>
<li><p><code><a href='pvalue.html'>pvalue</a></code></p></li>
</ul>

    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><a href='http://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12471/full'>http://onlinelibrary.wiley.com/doi/10.1111/1755-0998.12471/full</a>
Caye, Kevin et al. (2016) Fast Inference of Individual Admixture Coefficients Using Geographic Data. bioRxiv
doi:10.1101/080291. <a href='http://biorxiv.org/content/early/2016/10/12/080291'>http://biorxiv.org/content/early/2016/10/12/080291</a></p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='tess3Main.html'>tess3Main</a></code>, <code><a href='plot.tess3Q.html'>plot.tess3Q</a></code>,
<code><a href='barplot.tess3Q.html'>barplot.tess3Q</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>tess3r</span>)

<span class='co'># Arabidopsis thaliana data set</span>
<span class='fu'>data</span>(<span class='no'>data.at</span>)
<span class='no'>genotype</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>X</span>
<span class='no'>coordinates</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>coord</span>

<span class='co'># Running the tess3 function</span>
<span class='no'>tess3.obj</span> <span class='kw'>&lt;-</span> <span class='fu'>tess3</span>(<span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>genotype</span>, <span class='kw'>coord</span> <span class='kw'>=</span> <span class='no'>coordinates</span>, <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fl'>4</span>,
                   <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>,
                   <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>1</span>)</div><div class='output co'>#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done
#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done
#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done
#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done</div><div class='input'>
<span class='co'># Plot error</span>
<span class='fu'>plot</span>(<span class='no'>tess3.obj</span>, <span class='kw'>pch</span> <span class='kw'>=</span> <span class='fl'>19</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"blue"</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Number of ancestral populations"</span>,
     <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Cross-validation score"</span>)</div><div class='img'><img src='tess3-1.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
<span class='co'># Retrieve the Q-matrix for K = 3 clusters</span>
<span class='no'>q.matrix</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='qmatrix.html'>qmatrix</a></span>(<span class='no'>tess3.obj</span>, <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>)

<span class='co'>## STRUCTURE-like barplot for the Q-matrix</span>
<span class='fu'>barplot</span>(<span class='no'>q.matrix</span>, <span class='kw'>border</span> <span class='kw'>=</span> <span class='fl'>NA</span>, <span class='kw'>space</span> <span class='kw'>=</span> <span class='fl'>0</span>,
       <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Individuals"</span>, <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Ancestry proportions"</span>,
       <span class='kw'>main</span> <span class='kw'>=</span> <span class='st'>"Ancestry matrix"</span>) <span class='kw'>-&gt;</span> <span class='no'>bp</span></div><div class='output co'>#&gt; Use CreatePalette() to define color palettes.</div><div class='input'><span class='fu'>axis</span>(<span class='fl'>1</span>, <span class='kw'>at</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fu'>nrow</span>(<span class='no'>q.matrix</span>), <span class='kw'>labels</span> <span class='kw'>=</span> <span class='no'>bp</span>$<span class='no'>order</span>, <span class='kw'>las</span> <span class='kw'>=</span> <span class='fl'>3</span>, <span class='kw'>cex.axis</span> <span class='kw'>=</span> <span class='fl'>.4</span>)</div><div class='img'><img src='tess3-2.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
<span class='co'>## Spatial interpolation of ancestry coefficient</span>
<span class='no'>my.colors</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(<span class='st'>"tomato"</span>, <span class='st'>"orange"</span>, <span class='st'>"lightblue"</span>)
<span class='no'>my.palette</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='CreatePalette.html'>CreatePalette</a></span>(<span class='no'>my.colors</span>, <span class='fl'>9</span>)
<span class='fu'>plot</span>(<span class='no'>q.matrix</span>, <span class='no'>coordinates</span>, <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"map.max"</span>, <span class='kw'>interpol</span> <span class='kw'>=</span> <span class='fu'><a href='FieldsKrigModel.html'>FieldsKrigModel</a></span>(<span class='fl'>10</span>),
     <span class='kw'>main</span> <span class='kw'>=</span> <span class='st'>"Ancestry coefficients"</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Longitude"</span>, <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Latitude"</span>,
     <span class='kw'>resolution</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>500</span>,<span class='fl'>500</span>), <span class='kw'>cex</span> <span class='kw'>=</span> <span class='fl'>.4</span>,
     <span class='kw'>col.palette</span> <span class='kw'>=</span> <span class='no'>my.palette</span>)</div><div class='output co'>#&gt; <span class='message'>This function required to attach maps namespace.</span></div><div class='img'><img src='tess3-3.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
<span class='co'>## Genome scan p-values for K = 3</span>
<span class='no'>p.values</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='pvalue.html'>pvalue</a></span>(<span class='no'>tess3.obj</span>, <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>)
<span class='fu'>hist</span>(<span class='no'>p.values</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"lightblue"</span>)</div><div class='img'><img src='tess3-4.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
<span class='co'>## Manhatan plot</span>
<span class='fu'>plot</span>(<span class='no'>p.values</span>, <span class='kw'>main</span> <span class='kw'>=</span> <span class='st'>"Manhattan plot"</span>,
    <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Locus id"</span>,
    <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"-log10(P-values)"</span>,
    <span class='kw'>cex</span> <span class='kw'>=</span> <span class='fl'>.3</span>, <span class='kw'>col</span> <span class='kw'>=</span> <span class='st'>"grey"</span>)</div><div class='img'><img src='tess3-5.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Kevin Caye, Olivier Francois.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
