<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Estimate ancestry coefficients and run genome scans for selection — tess3Main • tess3r</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">tess3r</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">Home</a>
</li>
<li>
  <a href="../articles/main-vignette.html">Overview</a>
</li>
<li>
  <a href="../reference/index.html">Documentation</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bcm-uga/TESS3_encho_sen">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate ancestry coefficients and run genome scans for selection</h1>
    </div>

    
    <p><code>tess3Main</code> estimates spatial population structure using a graph based non
negative matrix factorization. After estimating the population structure is used to
compute a Fst statistic for each locus. See references for more details.</p>
    

    <pre class="usage"><span class='fu'>tess3Main</span>(<span class='no'>X</span>, <span class='kw'>XProba</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='no'>coord</span>, <span class='no'>K</span>, <span class='no'>ploidy</span>, <span class='kw'>lambda</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>W</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>, <span class='kw'>max.iteration</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>tolerance</span> <span class='kw'>=</span> <span class='fl'>1e-05</span>,
  <span class='kw'>openMP.core.num</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>Q.init</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>mask</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>copy</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>algo.copy</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>X</th>
      <td><p>a numeric matrix which corresponds to the genotype matrix. This matrix
must be of size \(n \times L\) where \(n\) is the number of individuals and
\(L\) is the number of loci. Values of this matrix are integers corresponding
to the number of variant alleles observed at a locus. If <code>NULL</code>, <code>XProba</code>
is used.</p></td>
    </tr>
    <tr>
      <th>XProba</th>
      <td><p>a numeric matrix which corresponds to genotype likelihoods (probabilities).
This matrix must be of size \(n \times (ploidy + 1)L\) where
\(n\) is the number of individuals and \(L\) is the number of loci. Entries of
this matrix are numeric values between 0 and 1 corresponding
to genotype probability. If <code>NULL</code>, this matrix is computed from the genotype matrix <code>X</code>.
See reference for details.</p></td>
    </tr>
    <tr>
      <th>coord</th>
      <td><p>a numeric matrix of size \(n \times 2\) where \(n\) is the
number of individuals. It contains the geographic coordinates (Longitude, Latitude) of
all sampled individuals.</p></td>
    </tr>
    <tr>
      <th>K</th>
      <td><p>an integer corresponding to
the number of ancestral populations.</p></td>
    </tr>
    <tr>
      <th>ploidy</th>
      <td><p>an integer which corresponds to the ploidy of the number of copy of chromosomes.</p></td>
    </tr>
    <tr>
      <th>lambda</th>
      <td><p>a nonnegative numeric which corresponds to the
spatial regularization parameter.</p></td>
    </tr>
    <tr>
      <th>W</th>
      <td><p>a numeric matrix which corresponds to the graph weight matrix.
If <code>NULL</code>, W is computed as
<code>W[i,j] = exp(-(coord[i] - coord[j])^2 / sigma^2)</code>,
where <code>coord[i]</code> is the set of geographic coordinates for individual i and
<code>sigma</code> equals 5 percent of the average geographic distance between individuals.</p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p><code>"projected.ls"</code> or <code>"qp"</code>. If <code>"projected.ls"</code>,
an alternating projected least squares algorithm is used. If <code>"qp"</code>,
an alternating quadratic programing algorithm is used. See references for details.</p></td>
    </tr>
    <tr>
      <th>max.iteration</th>
      <td><p>the maximum number of
iterations in the optimization algorithm.</p></td>
    </tr>
    <tr>
      <th>tolerance</th>
      <td><p>a numeric value which corresponds to the tolerance paramter in the
stopping criterion of the optimization algorithm.</p></td>
    </tr>
    <tr>
      <th>openMP.core.num</th>
      <td><p>number of core used by the algorithm. It requires that openMP is
installed.</p></td>
    </tr>
    <tr>
      <th>Q.init</th>
      <td><p>a numeric matrix which corresponds to the initial value of
<code>Q</code> for the algorithm.</p></td>
    </tr>
    <tr>
      <th>mask</th>
      <td><p>if not <code>NULL</code>, this numeric value is the proportion of genotypic matrix entries
which are masked when computing the cross validation criterion.</p></td>
    </tr>
    <tr>
      <th>copy</th>
      <td><p>if TRUE data will be copied once.</p></td>
    </tr>
    <tr>
      <th>algo.copy</th>
      <td><p>if TRUE, data will be copied in order to speed the algorithm.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>If <code>TRUE</code> run information is printed.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of class tess3Main which is a list with the following attributes:</p><dl class='dl-horizontal'>
   <dt>L</dt><dd><p>the number of loci</p></dd>
   <dt>n</dt><dd><p>the number of individuals</p></dd>
   <dt>ploidy</dt><dd><p>the number of copies of chromosomes.</p></dd>
   <dt>K</dt><dd><p>the number of ancestral populations.</p></dd>
   <dt>G</dt><dd><p>the ancestral genotype frequency matrix.</p></dd>
   <dt>Q</dt><dd><p>the ancestry coefficient matrix.</p></dd>
   <dt>Fst</dt><dd><p>Fst statistic computed at each locus.</p></dd>
   <dt>Fscore</dt><dd><p>Fscores computed from the Fst statistics.</p></dd>
   <dt>pvalue</dt><dd><p>pvalues computed from the Fscores.</p></dd>
   <dt>log.pvalue</dt><dd><p>The \(log(pvalue)\).</p></dd>
   <dt>rmse</dt><dd><p>root square mean error between <code>XProba</code> and
   <code>tcrossprod(Q, G)</code>.</p></dd>
   <dt>crossentropy</dt><dd><p>cross-entropy error between <code>XProba</code> and
   <code>tcrossprod(Q, G)</code>.</p></dd>
  <dt>crossvalid.rmse</dt><dd><p>if masked is not <code>NULL</code>, root square mean error
  between <code>XProba[masked]</code> and <code>tcrossprod(Q, G)[masked]</code>.</p></dd>
   <dt>crossvalid.crossentropy</dt><dd><p>if masked not <code>NULL</code>,
   the cross-entropy error between <code>XProba[masked]</code> and
   <code>tcrossprod(Q, G)[masked]</code>.</p></dd>
</dl>

    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><a href='https://hal.archives-ouvertes.fr/hal-01222555/'>https://hal.archives-ouvertes.fr/hal-01222555/</a> <a href='http://biorxiv.org/content/early/2016/10/12/080291'>http://biorxiv.org/content/early/2016/10/12/080291</a></p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='tess3.html'>tess3</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>library</span>(<span class='no'>tess3r</span>)

<span class='co'># Arabidopsis thaliana data set</span>
<span class='fu'>data</span>(<span class='no'>data.at</span>)
<span class='no'>genotype</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>X</span>
<span class='no'>coordinates</span> <span class='kw'>&lt;-</span> <span class='no'>data.at</span>$<span class='no'>coord</span>

<span class='co'># Run of tess3 main algorithm</span>
<span class='no'>tess3.obj</span> <span class='kw'>&lt;-</span> <span class='fu'>tess3Main</span>(<span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>genotype</span>,
                      <span class='kw'>coord</span> <span class='kw'>=</span> <span class='no'>coordinates</span>,
                      <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>,
                      <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>,
                      <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>1</span>)</div><div class='output co'>#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done</div><div class='input'>
<span class='co'># Run of tess3 main algorithm with cross validation errors computation.</span>
<span class='no'>tess3.obj</span> <span class='kw'>&lt;-</span> <span class='fu'>tess3Main</span>(<span class='kw'>X</span> <span class='kw'>=</span> <span class='no'>genotype</span>,
                      <span class='kw'>coord</span> <span class='kw'>=</span> <span class='no'>coordinates</span>,
                      <span class='kw'>K</span> <span class='kw'>=</span> <span class='fl'>3</span>,
                      <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"projected.ls"</span>,
                      <span class='kw'>ploidy</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                      <span class='kw'>mask</span> <span class='kw'>=</span> <span class='fl'>0.05</span>)</div><div class='output co'>#&gt; == Computing spectral decomposition of graph laplacian matrix: done
#&gt; ==Main loop with 1 threads: done</div><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Kevin Caye, Olivier Francois.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
